pkg/router/controller/router_controller.go

```diff
@@ -72,64 +59,9 @@ func (c *RouterController) Run() {
 		c.HandleNamespaces()
 		go utilwait.Forever(c.HandleNamespaces, c.NamespaceSyncInterval)
 	}
-	go utilwait.Forever(c.HandleRoute, 0)
-	go utilwait.Forever(c.HandleEndpoints, 0)
-	if c.WatchNodes {
-		go utilwait.Forever(c.HandleNode, 0)
-	}
-	if c.EnableIngress {
-		go utilwait.Forever(c.HandleIngress, 0)
-		go utilwait.Forever(c.HandleSecret, 0)
-	}
-	go c.watchForFirstSync()
-}
-
-// handleFirstSync signals the router when it sees that the various
-// watchers have successfully listed data from the api.
-func (c *RouterController) handleFirstSync() bool {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	synced := c.RoutesListSuccessfulAtLeastOnce() &&
-		c.EndpointsListSuccessfulAtLeastOnce() &&
-		(c.Namespaces == nil || c.filteredByNamespace) &&
-		(!c.EnableIngress ||
-			(c.IngressesListSuccessfulAtLeastOnce() && c.SecretsListSuccessfulAtLeastOnce()))
-	if !synced {
-		return false
-	}
-
-	// If any of the event queues were empty after the initial List,
-	// the tracking listConsumed variable's default value of 'false'
-	// may prevent the router from committing.  Set the value to
-	// 'true' to ensure that state can be committed if necessary.
-	if c.RoutesListCount() == 0 {
-		c.routesListConsumed = true
-	}
-	if c.EndpointsListCount() == 0 {
-		c.endpointsListConsumed = true
-	}
-	if c.EnableIngress {
-		if c.IngressesListCount() == 0 {
-			c.ingressesListConsumed = true
-		}
-		if c.SecretsListCount() == 0 {
-			c.secretsListConsumed = true
-		}
-	}
-	c.commit()
-
-	return true
-}
-
-// watchForFirstSync loops until the first sync has been handled.
-func (c *RouterController) watchForFirstSync() {
-	for {
-		if c.handleFirstSync() {
-			return
-		}
-		time.Sleep(50 * time.Millisecond)
-	}
+	c.StartInformers(utilwait.NeverStop)
```

Comment:
This doesn't seem to belong here, but in the factory.  Why should this be a concern of the route controller?

Registering functions into another type is a code smell.  Just have a higher level method that calls public methods on the controller from the factory